### DISCUSSION TITLE
About `_textEditingController`

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#141] [General]
Is there any particular reason to use `memo` instead `value` here?:
```dart
TextEditingController _textEditingController(
  WidgetSideEffectRegistrar use, {
  String? initialText,
}) {
  final controller = use.memo(() => TextEditingController(text: initialText));
  use.effect(() => controller.dispose, [controller]);
  return controller;
}
```

I'm just about start implementing a ReArch port to `Dart ReAct` and I just want to use the best tools in the best way.

**Comment by busslina** [comment]
I just started it, but don't know when I will finish it. Any advice or contribution is welcome. I hope that it won't be "extremely hard" for me haha.

https://github.com/busslina/rearch-dart/tree/introducing-react-rearch/packages/react_rearch

**Reply by busslina** [reply]
Wow... Is more complex to do that I thought.
- Dart React is way different to Flutter Widget
- I don't know about Flutter Widget internals (just a 5%)
- I have to understand that Flutter Widget internals and how flutter_rearch interacts with it before creating an "analogous" implementation.

Aborting for now haha

**Reply by GregoryConrad** [reply]
At a really high level, the widget tree is turned into a corresponding Element tree that flutter uses to actually render everything. flutter_rearch uses the RearchConsumer to build a custom element type that can interop between the Element/Flutter side and ReArch's CapsuleContainers.

**Reply by busslina** [reply]
Yeah, so what I couldn't see is that I need to (re)create a React Component Tree, which I don't know if it is possible to do, because Dart React is just a wrapper on JS React, so seems difficult... What do you think?

**Reply by busslina** [reply]
Maybe the easier solution for me is to go back to Flutter Web and try to minimize the most the `dart.main.js` file size.

**Reply by busslina** [reply]
I discovered this other package [over_react](https://pub.dev/packages/over_react) which I will study to see if it can provide a sort of Widget Tree

**Reply by GregoryConrad** [reply]
You should be able to get away with just calls to `setState` or similar if you manually define your own custom component, I'd think

**Reply by busslina** [reply]
> You should be able to get away with just calls to `setState` or similar if you manually define your own custom component, I'd think

Components are defined and registered, and only the registered objects are suppose to be use directly on building (like the built-in objects div, label, ...). So seems difficult because I can add a `setState` method, but the registered object would't care about it.

```dart
var myComponent = registerComponent2(() => _MyComponent ());

class _MyComponent extends Component2 {
[...]
}
```

So another alternative I thought is to build my own framework on top of [web package](https://pub.dev/packages/web) and ReArch. Yes, I'm a little bit crazy haha

UPDATE:
... or fork the `react` package to add that `setState` method

**Reply by GregoryConrad** [reply]
I thought when you make custom class-based components in React you get a `setState` method (or something similar) that you can call to update a component: https://legacy.reactjs.org/docs/react-component.html#setstate

You can integrate that with `CapsuleContainer.onNextUpdate` to get your components to receive container updates.

**Reply by busslina** [reply]
Okay, gonna see...

`[...]` includes all that functions (component lifecycle + setstate + ...)

**Reply by busslina** [reply]
Done. Gonna test it: https://github.com/busslina/rearch-dart/blob/introducing-react-rearch/packages/react_rearch/lib/src/widgets/consumer.dart

**Reply by busslina** [reply]
Yessss, seems working. It has to contain any bug somewhere hehe.
![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/4c07ea8d-5f83-4075-af67-d10d0df69a3a)
![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/83692b80-1848-4cec-beac-3cccb8d3e158)

I will test it more migrating a small/medium project.

**Reply by busslina** [reply]
https://github.com/busslina/rearch-dart/tree/introducing-react-rearch/packages/react_rearch/example

**Reply by GregoryConrad** [reply]
Pretty neat!

**Comment by GregoryConrad** [comment]
The reason for memo over value here is mostly style, but memo is often used in situations like this when the value it returns _could_ change in the future. In the text editing controller example, it won't since initialText is irrelevant after it has been created, but it many other situations a change to the initialText (or whatever the constructor parameter may be) would warrant recreating the object.
